packages/peregrine/src/Toasts/useToastContext.js

```diff
@@ -0,0 +1,127 @@
+import React, { createContext, useContext, useReducer } from 'react';
+/**
+ * The state of the toast store.
+ * @typedef {Object} ToastState
+ * @property {Object} toasts - unordered map of id:toast
```

Comment:
I was under the impression that objects/immutables are preferred for state values. Using a map for `toasts` would prevent me from being able to simply spread the previous state or the new state. I'd probably have to copy the map from the previous state, modify the toast, set that value in the copy, and then assign the copy to the `toasts` object in state. Is there a simpler way that I'm missing?

Reply:
> I was under the impression that objects/immutables are preferred for state values.

Redux prefers plain objects because they're serializable and compatible with its devtool extensions. Also, a Redux state tree tends to be very deep and gets sliced a lot. I don't think that strictness is quite as important for local state reducers, especially when maps are more suitable under the conditions I listed above.

To be clear, neither a plain object nor a `Map` are immutable, we just treat them as if they were.

```es6
const initialState = {
  toasts: new Map()
}

const reducer = (state, action = {}) => {
  const { payload, type } = action

  switch (type) {
    case "remove": {
      // a map is iterable, and each entry is a `[key, value]` array
      // which happens to be the shape expected by its constructor
      // so you can always call `new Map(currentMap)`
      const nextToasts = new Map(state.toasts)

      // the `delete` method returns a boolean
      // `true` if an entry was deleted, `false` if not
      // but it won't throw if it doesn't find an entry
      nextToasts.delete(payload)

      // we still spread the state object
      // since we only used a map for the toasts
      return {
        ...state,
        toasts: nextToasts
      }
    }
  }
}
```